include/analyzers/analyzer.h

```diff
@@ -48,71 +69,88 @@ class analyzer
 
     /**
      * Tokenizes a document.
-     * @param doc The document to store the tokenized information in
+     * @param doc The document to be tokenized
+     * @return a feature_map that maps the observed features to their
+     *  counts in the document
      */
-    virtual void tokenize(corpus::document& doc) = 0;
+    feature_map analyze(const corpus::document& doc);
 
     /**
      * Clones this analyzer.
      */
     virtual std::unique_ptr<analyzer> clone() const = 0;
 
-    /**
-     * @param config The config group used to create the analyzer from
-     * @return an analyzer as specified by a config object
-     */
-    static std::unique_ptr<analyzer> load(const cpptoml::table& config);
+    friend multi_analyzer<T>;
 
+  private:
     /**
-     * @param config The config group used to create the analyzer from
-     * @return the default filter chain for this version of MeTA,
-     * based on a config object
+     * The tokenization function that actually does the heavy lifting. This
+     * should be overridden in derived classes.
+     *
+     * @param doc The document to be tokenized
+     * @param counts The feature_map to place observed feature counts into
      */
-    static std::unique_ptr<token_stream>
-        default_filter_chain(const cpptoml::table& config);
+    virtual void tokenize(const corpus::document& doc, feature_map& counts) = 0;
+};
 
-    /**
-     * @param config The config group used to create the analyzer from
-     * @return the default filter chain for unigram words for this version
-     * of MeTA, based on a config object
-     */
-    static std::unique_ptr<token_stream>
-        default_unigram_chain(const cpptoml::table& config);
+/**
+ * Basic exception for analyzer interactions.
+ */
+class analyzer_exception : public std::runtime_error
+{
+  public:
+    using std::runtime_error::runtime_error;
+};
 
-    /**
-     * @param global The original config object with all parameters
-     * @param config The config group used to create the filters from
-     * @return a filter chain as specified by a config object
-     */
-    static std::unique_ptr<token_stream>
-        load_filters(const cpptoml::table& global,
-                     const cpptoml::table& config);
+/**
+ * @param config The config group used to create the analyzer from
+ * @return an analyzer as specified by a config object
+ */
+template <class T>
+std::unique_ptr<analyzer<T>> load(const cpptoml::table& config);
 
-    /**
-     * @param src The token stream that will feed into this filter
-     * @param config The config group used to create the filter from
-     * @return a single filter specified by a config object
-     */
-    static std::unique_ptr<token_stream>
-        load_filter(std::unique_ptr<token_stream> src,
-                    const cpptoml::table& config);
+extern template std::unique_ptr<analyzer<uint64_t>>
```

Comment:
So these show up all over the place now. Basically every `analyzer` header file and every `tree_featurizer` header file. Do you want me to document it everywhere, or just here?

Essentially what this is doing is telling the compiler to not attempt to instantiate the template for the specified type in all translation units that include this header. This would obviously cause an error at link time unless you link with an object file that actually has the instantiation. We force the compiler to instantiate the templates *once* in the corresponding `.cpp` file. This is what saves us from having to move everything to be header-only and parse and instantiate every function that's used in every translation unit.

Hmm... I guess an easier explanation is that this:

```cpp
extern template class myanalyzer<uint64_t>;
```

is the declaration and 

```cpp
template class myanalyzer<uint64_t>;
```

is the definition, to use the standard terms for regular classes. See: http://en.cppreference.com/w/cpp/language/class_template#Explicit_instantiation

**tl;dr:** Not sure how you want me to document what this is doing, beyond something like


```cpp
// declare the valid instantiations for this analyzer
extern template class ngram_word_analyzer<uint64_t>;
extern template class ngram_word_analyzer<double>;

// declare the valid instantiations for this analyzer's trait class
extern template struct analyzer_traits<ngram_word_analyzer<uint64_t>>;
extern template struct analyzer_traits<ngram_word_analyzer<double>>;
```